clippy_lints/src/undocumented_unsafe_block_safety.rs

```diff
@@ -0,0 +1,198 @@
+    None,
+}
+
+impl LateLintPass<'_> for UndocumentedUnsafeBlockSafety {
+    fn check_block(&mut self, cx: &LateContext<'_>, block: &'_ Block<'_>) {
+        if self.local {
+            self.local = false;
+            return
+        }
+
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCK_SAFETY, block.hir_id);
+            if !in_external_macro(cx.tcx.sess, block.span);
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                find_candidate(cx, block.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+
+    fn check_local(&mut self, cx: &LateContext<'_>, local: &'_ Local<'_>) {
+        if_chain! {
+            if !is_lint_allowed(cx, UNDOCUMENTED_UNSAFE_BLOCK_SAFETY, local.hir_id);
+            if !in_external_macro(cx.tcx.sess, local.span);
+            if let Some(init) = local.init;
+            if let ExprKind::Block(block, _) = init.kind;
+            if let BlockCheckMode::UnsafeBlock(UnsafeSource::UserProvided) = block.rules;
+            if let Some(enclosing_scope_hir_id) = cx.tcx.hir().get_enclosing_scope(block.hir_id);
+            then {
+                self.local = true;
+                find_candidate(cx, local.span, enclosing_scope_hir_id)
+            }
+        }
+    }
+}
+
+fn find_candidate(cx: &LateContext<'_>, span: Span, enclosing_hir_id: HirId) {
+    if let Some(true) = block_has_safety_comment(cx.tcx, enclosing_hir_id, span) {
+        return;
+    }
+
+    let block_indent = indent_of(cx, span);
+    let suggestion = format!("// Safety: ...\n{}", snippet(cx, span, ".."));
+
+    span_lint_and_sugg(
+        cx,
+        UNDOCUMENTED_UNSAFE_BLOCK_SAFETY,
+        span,
+        "unsafe block missing a safety comment",
+        "consider adding a safety comment",
+        reindent_multiline(Cow::Borrowed(&suggestion), true, block_indent).to_string(),
+        Applicability::HasPlaceholders,
+    );
+}
+
+fn block_has_safety_comment(tcx: TyCtxt<'_>, enclosing_hir_id: HirId, block_span: Span) -> Option<bool> {
+    let map = tcx.hir();
+    let source_map = tcx.sess.source_map();
+
+    let enclosing_scope_span = map.opt_span(enclosing_hir_id)?;
+    let between_span = enclosing_scope_span.until(block_span);
+
+    let file_name = source_map.span_to_filename(between_span);
+    let source_file = source_map.get_source_file(&file_name)?;
+
+    let lex_start = (between_span.lo().0 + 1) as usize;
+    let mut src_str = source_file.src.as_ref()?[lex_start..between_span.hi().0 as usize].to_string();
+
+    // Remove all whitespace, but retain newlines to verify the block immediately follows the comment
+    src_str.retain(|c| c == '\n' || !c.is_whitespace());
+
+    let mut src_str_split = src_str.lines().rev();
+
+    let mut found_safety_comment = false;
+
+    while let Some(line) =  src_str_split.next() {
```

Comment:
I think using the Rust lexer here is the way to go, like #7557 did. You already do the span manipulation, before re-lexing the code (by hand), so you already addressed my comment https://github.com/rust-lang/rust-clippy/pull/7557/files#r687488132